mods/key_rev_value.js

```diff
@@ -0,0 +1,295 @@
+    var rp = req.params;
+    var storeReq = {
+        uri: new URI([rp.domain,'sys','table',rp.bucket,'']),
+        body: {
+            table: rp.bucket,
+            attributes: {
+                key: rp.key
+            },
+            limit: 1
+        }
+    };
+    if (rp.revision) {
+        storeReq.body.attributes.rev = parseRevision(rp.revision);
+        if (rp.tid) {
+            storeReq.body.attributes.tid = coerceTid(rp.tid);
+        }
+    }
+    return restbase.get(storeReq).then(returnRevision(req));
+};
+
+
+KRVBucket.prototype.listRevisions = function(restbase, req) {
+    var rp = req.params;
+    var storeRequest = {
+        uri: new URI([rp.domain,'sys','table',rp.bucket,'']),
+        body: {
+            table: req.params.bucket,
+            attributes: {
+                key: req.params.key
+            },
+            proj: ['rev', 'tid']
+        }
+    };
+    if (rp.revision) {
+        storeRequest.body.attributes.rev = parseRevision(rp.revision);
+    }
+    return restbase.get(storeRequest)
+    .then(function(res) {
+        return {
+            status: 200,
+            headers: {
+                'content-type': 'application/json'
+            },
+            body: {
+                items: res.body.items.map(function(row) {
+                    return { revision: row.rev, tid: row.tid };
+                })
+            }
+        };
+    });
+};
+
+
+KRVBucket.prototype.putRevision = function(restbase, req) {
+    var rp = req.params;
+    var rev = parseRevision(rp.revision);
+    var tid = rp.tid && coerceTid(rp.tid) || uuid.v1();
+    if (req.headers['last-modified']) {
+        // XXX: require elevated rights for passing in the revision time
+        tid = rbUtil.tidFromDate(req.headers['last-modified']);
+    }
+
+    var storeReq = {
+        uri: new URI([rp.domain,'sys','table',rp.bucket,'']),
+        body: {
+            table: rp.bucket,
+            attributes: {
+                key: rp.key,
+                rev: rev,
+                tid: tid,
+                value: req.body,
+                'content-type': req.headers['content-type']
+                // TODO: include other data!
+            }
+        }
+    };
+    return restbase.put(storeReq)
+    .then(function(res) {
+        if (res.status === 201) {
+            return {
+                status: 201,
+                headers: {
+                    etag: rp.revision
+                },
+                body: {
+                    message: "Created.",
+                    tid: rp.revision
+                }
+            };
+        } else {
+            throw res;
+        }
+    })
+    .catch(function(error) {
+        restbase.log('error/kv/putRevision', error);
+        return { status: 400 };
+    });
+};
+
+module.exports = function(options) {
+    var krvBucket = new KRVBucket(options);
+
+    return {
+        spec: spec, // Re-export from spec module
+        test: null, // Spec test function from spec module
```

Comment:
Is the optional `test` function meant for tests that are hard to write in terms of specs and `x-ample`s?